corehq/apps/sms/management/commands/populate_app_id_for_sms.py

```diff
@@ -0,0 +1,51 @@
+import logging
+
+from django.core.management.base import BaseCommand
+
+from corehq.apps.app_manager.util import get_app_id_from_form_unique_id
+from corehq.apps.sms.models import Keyword, KeywordAction, MessagingEvent, MessagingSubEvent
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    help = """
+        Populate any KeywordAction, MessagingEvent, and MessagingSubEvent models
+        that contain a form_unique_id with the associated app_id.
+    """
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--dry-run',
+            action='store_true',
+            default=False,
+            help='Do not actually modify the database, just verbosely log what will happen',
+        )
+
+    def handle(self, dry_run=False, **options):
+        def _update_objects(model, domain, form_unique_id):
+            log_prefix = "{} Domain {}, {}, form unique_id {}".format("[DRY RUN]" if dry_run else "",
+                                                                      model.__name__,
+                                                                      domain,
+                                                                      form_unique_id)
+            app_id = get_app_id_from_form_unique_id(domain, form_unique_id)
+            if app_id:
+                to_update = model.objects.filter(form_unique_id=action.form_unique_id, app_id__isnull=True)
+                if not dry_run:
+                    to_update.update(app_id=app_id)
+                logger.info("{}: Updated {} models to use app id {}".format(log_prefix, to_update.count(), app_id))
+            else:
+                logger.info("{}: Could not find app".format(log_prefix))
+
+        for domain_keyword in Keyword.objects.distinct('domain'):
```

Comment:
This fixes the bug. Compare these three loops to the two here: https://github.com/dimagi/commcare-hq/blob/d362bb195955e35be50ff9a320e95815c79cd619/corehq/apps/sms/migrations/0037_app_id.py#L12-L13

The old code grabbed one keyword per domain and only updated actions for that keyword, instead of going through all keywords in the domain.

Reply:
Apologies if i am missing something obvious here,
why are we filtering by `domain` anyway if we are iterating all Keyword objects?

Reply:
Django's `QuerySet.distinct()` is a weird animal with a lot of caveats. I think this query might be slightly easier to understand and also more efficient if written as
```py
for domain, in Keyword.objects.values('domain').distinct('domain'):
    for keyword in Keyword.objects.filter(domain=domain):
```

The meaning of the first query is _get the distinct set of domains that have keyword objects_. However, I also wonder about @mkangia's question. Would this achieve the same result without the nested loops?
```py
from django.db.models.functions import Coalesce
actions = (
    KeywordAction.objects
    .filter(form_unique_id__isnull=False, app_id__isnull=True)
    .annotate(domain=Coalesce('keyword__domain', None))
)
for action in actions:
    _update_objects(KeywordAction, action.domain, action.form_unique_id)
```
The `Coalesce` does nothing, but is necessary to keep Django from freaking out because it can't do obvious things.

Reply:
@millerdev How strongly do you feel about that as an improvement? From my reading of the `annotate` docs I think the new code would be equivalent, but I don't know the ORM methods especially well. But I'm inclined to want to stop tinkering with this code so I can run the migration (SMS surveys are currently broken for some projects).

Reply:
I think the code without the nested loops (second code block in my comment) is far more efficient. However, will let it up to you to decide whether to use it or not. It may not matter if this will only be run a handful of times before it is obsolete.

Reply:
Okay. I'm going to go ahead and merge this - there are few enough SurveyKeyword objects that I'm not concerned about performance.